### PR TITLE
bpo-40129: Add fake number classes in test.support.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue40129

### CODE REVIEW

File: Lib/test/support/__init__.py [hunk 1]
```diff
@@ -3189,22 +3190,37 @@ def with_pymalloc():
...
+    __int__ = Fake._return
+
+class FakeFloat(Fake):
+    __float__ = Fake._return
+
+class FakeComplex(Fake):
+    __complex__ = Fake._return
+
+class FakePath(Fake):
+    """Simple implementing of the path protocol.
```
Comment (pganssle) [inline Lib/test/support/__init__.py:3221]:
```suggestion
    """Simple implementation of the path protocol.
```

File: Lib/test/support/__init__.py [hunk 2]
```diff
@@ -3189,22 +3190,37 @@ def with_pymalloc():
...
+    """
```
Comment (pganssle) [inline Lib/test/support/__init__.py:3222]: Should this be on the previous line? Looks weird to have it sitting on its own following a short one-line docstring like this.

**Review by pganssle — COMMENTED** [review]
Thanks for doing this. The idea seems sound and the datetime implementation seems right.

One suggestion: you should add documentation for this. There is [documentation for `test.support`](https://docs.python.org/3/library/test.html), but even a block comment before `Fake` would be helpful to give some additional context to future maintainers and devs about why it's there and when and where you should use it.

File: Lib/test/test_buffer.py [hunk 3]
```diff
@@ -2510,22 +2511,9 @@ def test_memoryview_sizeof(self):
...
-        class IDX(object):
-            def __init__(self, val):
-                self.val = val
-            def __index__(self):
-                return self.val
-
         def f(): return 7
 
-        values = [INT(9), IDX(9),
+        values = [FakeInt(9), FakeIndex(9),
```
Comment (skrah) [inline Lib/test/test_buffer.py:2516]: I don't quite understand the name, there's nothing really fake about these classes.  Also, it's nice to see what the classes do without leaving the file.

**Comment by skrah** [comment]
IMO tests should be primitive and locally understandable. If they get too fancy, it's easy to lose track of what is being tested.

So I'm not sure if much is gained here.

File: Lib/test/test_buffer.py (hunk 3 above)
Comment (mdickinson) [reply Lib/test/test_buffer.py:2516]: I agree with both points here.

When scanning through the PR on GitHub and seeing `FakeInt(10.9)`, my first thought is that I don't know what that means without guessing, so I want to quickly find the `FakeInt` definition.  That's harder when that definition is in another file. (And yes, if you have a decent editor open and set up properly, the definition should be just a click or keystroke away, but not everyone has that setup available all of the time.)

The problem could potentially be mitigated with a name that more clearly describes what the class is doing (`HasDunderIntReturning(3.2)`?), but as you can tell I'm having trouble coming up with a name that's going to be immediately unambiguous for people reading the code without also being unwieldy.

**Comment by serhiy-storchaka** [comment]
The main purpose is to make easier to test that the specified function supports an object that implements `__index__`, not just `int`. There are a lot of such tests in different files, and we can add more tests, but it is tiring to reimplement such class in every file. It would also help if all test classes used for this purpose have a standardized and well-recognized name, so you don't even need to search its definition.

I used name `FakeIndex` because we already have `FakePath` for similar purposes (used in 11 different files), and when it was introduced it looked a good name. We also have more specialized classes like `FakeSocket`. Do you have to propose better name? `MockIndex`, `IndexLike` or just `Index`?

**Pull request closed without merging**